File: scripts/fig4_overconfidence.py
"""Figure 4: Overconfidence scatter – actual vs self-estimated accuracy."""

import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib
matplotlib.use('Agg')
from sklearn.linear_model import LinearRegression

DATA_DIR = '/Users/mdisorbo/madm_colm/results/study3'
OUT_PATH = '/Users/mdisorbo/madm_colm/paper/figures/overconfidence_by_condition.png'

MODELS = [
    ('Qwen3.5-9B', 'Qwen3.5-9B'),
    ('Gemma 3 4B', 'gemma-3-4b-it'),
    ('Claude Sonnet 4.6', 'claude-sonnet-4-6'),
    ('Qwen3.5-397B', 'Qwen3.5-397B-A17B'),
    ('Gemma 3 12B', 'gemma-3-12b-it'),
    ('Claude Opus 4.7', 'claude-opus-4-7'),
]

DATASETS = {
    'HotelBookings': 'orange',
    'LendingClub': 'blue',
    'WikipediaToxicity': 'red',
    'MovieLens': 'purple',
}

ALL_DATASETS = ['HotelBookings', 'LendingClub', 'WikipediaToxicity', 'MovieLens']

fig, axes = plt.subplots(2, 3, figsize=(13, 8))
axes_flat = axes.flatten()

all_vals = []

# For each model: fit linear regression on hint data (pred_acc -> esc_rate),
# then invert to get self_estimated_acc from nohint esc_rate
model_results = {}
for short_name, tag in MODELS:
    # Load all hint data for this model
    hint_frames = []
    for ds in ALL_DATASETS:
        fpath = os.path.join(DATA_DIR, f'{ds}_summary_nothink_{tag}.csv')
        if os.path.exists(fpath):
            try:
                df = pd.read_csv(fpath)
            except Exception:
                continue
            if df.empty:
                continue
            df['dataset'] = ds
            hint_frames.append(df)
    if not hint_frames:
        continue
    hint_all = pd.concat(hint_frames, ignore_index=True)

    # Fit linear regression: esc_rate = slope * pred_acc + intercept
    X = hint_all['pred_acc'].values.reshape(-1, 1)
    y = hint_all['esc_rate'].values
    reg = LinearRegression().fit(X, y)
    slope = reg.coef_[0]
    intercept = reg.intercept_

    # For each dataset, load nohint data and compute self-estimated accuracy
    results = []
    for ds in ALL_DATASETS:
        nohint_path = os.path.join(DATA_DIR, f'{ds}_summary_nothink_nohint_{tag}.csv')
        hint_path = os.path.join(DATA_DIR, f'{ds}_summary_nothink_{tag}.csv')
        if not os.path.exists(nohint_path) or not os.path.exists(hint_path):
            continue
        try:
            nohint_df = pd.read_csv(nohint_path)
            hint_df = pd.read_csv(hint_path)
        except Exception:
            continue
        if nohint_df.empty or hint_df.empty:
            continue

        # Match conditions
        for _, row in nohint_df.iterrows():
            cond = row['condition']
            nohint_esc = row['esc_rate']
            # Invert: self_estimated_acc = (nohint_esc - intercept) / slope
            if slope != 0:
                self_est = np.clip((nohint_esc - intercept) / slope, 0, 1)
            else:
                self_est = np.nan
            # Get actual accuracy from hint data
            hint_row = hint_df[hint_df['condition'] == cond]
            if hint_row.empty:
                continue
            actual_acc = hint_row['pred_acc'].values[0]
            results.append({
                'dataset': ds,
                'condition': cond,
                'actual_acc': actual_acc,
                'self_est_acc': self_est,
            })
    if results:
        model_results[tag] = pd.DataFrame(results)
        all_vals.extend([r['actual_acc'] for r in results])
        all_vals.extend([r['self_est_acc'] for r in results])

vmin = 0.4
vmax = 1.0

for i, (short_name, tag) in enumerate(MODELS):
    ax = axes_flat[i]
    df = model_results.get(tag)
    if df is None:
        ax.set_visible(False)
        continue
    # Diagonal line
    ax.plot([vmin, vmax], [vmin, vmax], 'k--', alpha=0.5, linewidth=1)
    for ds, color in DATASETS.items():
        sub = df[df['dataset'] == ds]
        if sub.empty:
            continue
        label = ds if i == 0 else None
        ax.scatter(sub['actual_acc'], sub['self_est_acc'], c=color, s=50, alpha=0.7,
                   label=label, edgecolors='none')
    ax.set_xlim(vmin, vmax)
    ax.set_ylim(vmin, vmax)
    ax.set_aspect('equal', adjustable='box')
    ax.set_title(short_name, fontsize=16, fontweight='bold')
    ax.set_xlabel('Actual accuracy', fontsize=13)
    ax.set_ylabel('Self-estimated accuracy', fontsize=13)
    ax.tick_params(labelsize=12)


handles, labels = axes_flat[0].get_legend_handles_labels()
fig.legend(handles, labels, loc='lower center', ncol=4, fontsize=16, frameon=False)
plt.tight_layout(rect=[0, 0.07, 1, 1], h_pad=4.0)
fig.savefig(OUT_PATH, dpi=300, bbox_inches='tight')
print(f'Saved to {OUT_PATH}')

File: scripts/fig4_overconfidence.R
# Figure 4: Overconfidence scatter -- actual vs self-estimated accuracy (2x4 grid)
# Same-family pairs vertically aligned, small on top, large on bottom

library(tidyverse)

DATA_DIR <- "results/study3"

# Model definitions
models <- tribble(
  ~display,             ~tag,
  "Qwen3.5-9B",        "Qwen3.5-9B",
  "GPT-5-nano",        "gpt-5-nano",
  "Llama4-Maverick",   "Llama-4-Maverick-17B-128E-Instruct-FP8",
  "Mixtral-8x7B",      "Mixtral-8x7B-Instruct-v0.1",
  "Qwen3.5-397B",      "Qwen3.5-397B-A17B",
  "GPT-5-mini",        "gpt-5-mini",
  "Llama3.3-70B",      "Llama-3.3-70B-Instruct-Turbo",
  "Mistral-Small-24B", "Mistral-Small-24B-Instruct-2501"
)

datasets <- c("HotelBookings", "LendingClub", "WikipediaToxicity", "MovieLens")
ds_colors <- c(
  "HotelBookings" = "#E69F00",
  "LendingClub" = "#0072B2",
  "WikipediaToxicity" = "#D55E00",
  "MovieLens" = "#8B4DAB"
)

# ============================================================
# Load hint and nohint summaries
# ============================================================

load_summary <- function(tag, type = "hint") {
  pattern <- if (type == "hint") {
    paste0("_summary_nothink_", tag, ".csv")
  } else {
    paste0("_summary_nothink_nohint_", tag, ".csv")
  }
  map_dfr(datasets, function(ds) {
    fpath <- file.path(DATA_DIR, paste0(ds, pattern))
    if (file.exists(fpath)) {
      read_csv(fpath, show_col_types = FALSE) %>%
        filter(n > 0) %>%
        mutate(dataset = ds)
    }
  })
}

# ============================================================
# Compute self-estimated accuracy per model per condition
# ============================================================

overconf_all <- map_dfr(seq_len(nrow(models)), function(i) {
  tag <- models$tag[i]
  display <- models$display[i]

  hint <- load_summary(tag, "hint")
  nohint <- load_summary(tag, "nohint")
  if (nrow(hint) == 0 || nrow(nohint) == 0) return(NULL)

  # Fit regression on hint data: esc_rate ~ pred_acc
  fit <- lm(esc_rate ~ pred_acc, data = hint)
  slope <- coef(fit)["pred_acc"]
  intercept <- coef(fit)["(Intercept)"]

  # For each nohint condition, invert to get self-estimated accuracy
  nohint %>%
    inner_join(
      hint %>% select(condition, dataset, actual_acc = pred_acc),
      by = c("condition", "dataset")
    ) %>%
    mutate(
      self_est_acc = pmin(pmax((esc_rate - intercept) / slope, 0), 1),
      overconfident = self_est_acc > actual_acc,
      gap = self_est_acc - actual_acc,
      display = display
    )
})

# ============================================================
# Inspect the data
# ============================================================

# Summary per model
overconf_all %>%
  group_by(display) %>%
  summarise(
    n_conditions = n(),
    n_overconfident = sum(overconfident),
    pct_overconfident = round(mean(overconfident) * 100, 1),
    mean_gap = round(mean(gap) * 100, 1),
    mean_actual = round(mean(actual_acc) * 100, 1),
    mean_self_est = round(mean(self_est_acc) * 100, 1),
    .groups = "drop"
  ) %>%
  arrange(desc(pct_overconfident)) %>%
  print(n = Inf)

cat("\nTotal:", sum(overconf_all$overconfident), "/", nrow(overconf_all),
    "(", round(mean(overconf_all$overconfident) * 100), "%)\n")

# Per-model per-dataset breakdown
overconf_all %>%
  group_by(display, dataset) %>%
  summarise(
    n = n(),
    pct_overconf = round(mean(overconfident) * 100),
    mean_gap = round(mean(gap) * 100, 1),
    .groups = "drop"
  ) %>%
  print(n = Inf)

# Browse full data:
# View(overconf_all)

# Worst overconfidence cases:
# overconf_all %>% arrange(desc(gap)) %>% head(20) %>% View()

# Worst underconfidence cases:
# overconf_all %>% arrange(gap) %>% head(20) %>% View()

# ============================================================
# Load individual sample-level data (prompts, predictions, escalation decisions)
# ============================================================

load_samples <- function(model_tag, dataset, condition, nohint = FALSE) {
  hint_tag <- if (nohint) "_nothink_nohint_" else "_nothink_"
  fpath <- file.path(DATA_DIR, paste0(dataset, "_", condition, hint_tag, model_tag, ".csv"))
  if (file.exists(fpath)) {
    read_csv(fpath, show_col_types = FALSE) %>%
      mutate(dataset = dataset, model = model_tag, condition_name = condition, nohint = nohint)
  } else {
    cat("File not found:", fpath, "\n")
    NULL
  }
}

load_all_samples <- function(model_tag, dataset, nohint = FALSE) {
  hint_tag <- if (nohint) "_nothink_nohint_" else "_nothink_"
  summary_path <- file.path(DATA_DIR, paste0(dataset, "_summary", hint_tag, model_tag, ".csv"))
  if (!file.exists(summary_path)) return(NULL)
  conditions <- read_csv(summary_path, show_col_types = FALSE)$condition
  map_dfr(conditions, ~load_samples(model_tag, dataset, .x, nohint = nohint))
}

# Example: load nohint samples (what Figure 4 uses)
# all_lc_nohint <- load_all_samples("gpt-5-mini", "LendingClub", nohint = TRUE)
# View(all_lc_nohint)

# Compare hint vs nohint for same condition:
# hint <- load_samples("gpt-5-mini", "LendingClub", "fico_over_700", nohint = FALSE)
# nohint <- load_samples("gpt-5-mini", "LendingClub", "fico_over_700", nohint = TRUE)
# View(hint)
# View(nohint)

# Find the most overconfident conditions and inspect their nohint prompts:
# worst <- overconf_all %>% arrange(desc(gap)) %>% head(5)
# samples <- load_samples(worst$model[1], worst$dataset[1], worst$condition[1], nohint = TRUE)
# View(samples)

# ============================================================
# Plot
# ============================================================

display_order <- c(
  "Qwen3.5-9B", "GPT-5-nano", "Llama4-Maverick", "Mixtral-8x7B",
  "Qwen3.5-397B", "GPT-5-mini", "Llama3.3-70B", "Mistral-Small-24B"
)
overconf_all$display <- factor(overconf_all$display, levels = display_order)

ggplot(overconf_all, aes(x = actual_acc, y = self_est_acc, color = dataset)) +
  geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "gray50") +
  geom_point(size = 2, alpha = 0.7) +
  facet_wrap(~display, nrow = 2) +
  scale_color_manual(values = ds_colors) +
  coord_cartesian(xlim = c(0.4, 1), ylim = c(0.4, 1)) +
  labs(
    x = "Actual accuracy",
    y = "Self-estimated accuracy",
    color = "Dataset"
  ) +
  theme_minimal(base_size = 12) +
  theme(
    legend.position = "bottom",
    strip.text = element_text(face = "bold"),
    panel.grid.minor = element_blank()
  )

ggsave("paper/figures/fig4_R.png", width = 14, height = 7, dpi = 300)
cat("Saved to paper/figures/fig4_R.png\n")
